oauthenticator/cilogon.py

```diff
@@ -153,6 +163,13 @@ def authenticate(self, handler, data=None):
                 self.username_claim, sorted(resp_json.keys())
             )
             raise web.HTTPError(500, "Failed to get username from CILogon")
+
+        gotten_name, gotten_idp = username.split('@')
```

Comment:
Can we add one more conditional so that this whole block is inside `if self.idp_whitelist:`? I think there might be some cases (e.g. where `ePPN` is not used) that username doesn't have `@` in it, in which case this initial split will fail.

Reply:
Sure, that makes sense. However it might still fail if `if self.idp_whitelist`, no? 

So maybe it's better to put it in a `try` block, which on exception will set `gotten_idp` to either empty string or `None`. In this way we don't have strange exception if the split fails (it will still failin the `HTTPError` below).

What do you think>

Reply:
@minrk Just pinging you to make sure this does not fall in the oblivion. I'm waiting for your feedback on it. Thanks!

Reply:
Failing is okay, I think. If using an idp whitelist, it should be incumbent on the admin to choose a username field that will work, right? Or are there username fields where only some users will get names with `@` and not others?

We could also do:

```python
if '@' not in username:
    name = username
    idp = ''
else:
    name, idp = username.split('@', 1)
```

Reply:
Well, I don't know, since that is not something that I am in control of: the auth provider is. Maybe @jbasney has any insight? If @minrk think that is ok with failing, I'm ok with it too.

Reply:
The eppn and email claims should always have the user@domain form but the sub claim won't, and strictly speaking the sub claim is what clients should be using, so I agree it's good to put the split inside `if self.idp_whitelist`.

The idea of having multiple entries in `self.idp_whitelist` makes me nervous, because of the possibility of username collisions across multiple domains. You don't want jbasney@wisc.edu logging in to jbasney@illinois.edu's account. It seems safer to follow the example from google.py to allow only a single `self.hosted_domain` parameter when you have a single trusted domain.

Reply:
> The idea of having multiple entries in self.idp_whitelist

That's why stripping the domain is only allowed when `len(idp_whitelist) == 1`